api/src/main/kotlin/com/google/devtools/ksp/symbol/KSFunctionType.kt

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2020 Google LLC
+ * Copyright 2010-2020 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.devtools.ksp.symbol
+import com.google.devtools.ksp.processing.Resolver
+/**
+ * Holds the information for a [KSFunctionDeclaration] where type arguments are resolved as member
+ * of a specific [KSType].
+ *
+ * @see Resolver.asMemberOf
+ */
+interface KSFunctionType {
```

Comment:
Have you considered using the `Function0<R>`, `Function1<P1, R>`, `Function2<P1, P2, R>`, ... family? Seems to me that they could also serve the purpose. Currently, explicit function types like `val f: (int) -> int = ...` as well as lambdas are resolved to the `FunctionN` family in the rest of the API.

Reply:
Not sure what you mean, return `Function` from `asMemberOf`?
Can you link me some code where I can check how `FunctionN` is used?

I basically need a type that can hold return type, parameter types, receiver type and type parameters.

Reply:
I mean treating the function types as regular class types / `KSType`. In Kotlin/JVM, [function types]( https://github.com/JetBrains/kotlin/blob/master/spec-docs/function-types.md) are modeled in library and the `Function` family are normal interfaces. For example, the `KSType` of `x`'s type in  `val x: List<Int>.(Double) -> Float = ...` is `Function2<List<Int>, Double, Float>`, which is [defined as an interface](https://github.com/JetBrains/kotlin/blob/master/libraries/stdlib/jvm/runtime/kotlin/jvm/functions/Functions.kt) in Kotlin stdlib.

To get the corresponding `KotlinType`, you may try [`KotlinBuiltIns.getFunction()`](https://github.com/JetBrains/kotlin/blob/master/core/descriptors/src/org/jetbrains/kotlin/builtins/KotlinBuiltIns.java#L292)`.getDefaultType()`.

The instance of `KotlinBuiltIns` can be obtained like [this](https://github.com/google/ksp/blob/master/compiler-plugin/src/main/kotlin/com/google/devtools/ksp/processing/impl/ResolverImpl.kt#L418).

Although it is designed around JVM, I think the idea is also applicable to other platforms. We can provide implementations for other platforms when needed in the future. The benefit is that we get an unified `KSType` for both functions and regular classes.

Reply:
as discussed offline, renamed it to `KSFunction`